yjit/src/codegen.rs

```diff
@@ -3761,34 +3786,38 @@ fn jit_rb_str_concat(
     _argc: i32,
     _known_recv_class: *const VALUE,
 ) -> bool {
-    let comptime_arg = jit_peek_at_stack(jit, ctx, 0);
-    let comptime_arg_type = ctx.get_opnd_type(StackOpnd(0));
 
-    // String#<< can take an integer codepoint as an argument, but we don't optimise that.
-    // Also, a non-string argument would have to call .to_str on itself before being treated
-    // as a string, and that would require saving pc/sp, which we don't do here.
-    // TODO: figure out how we should optimise a string-subtype argument here
-    if comptime_arg_type != Type::CString && comptime_arg.class_of() != unsafe { rb_cString } {
+    // The << operator can accept integer codepoints for characters
+    // as the argument. We only specially optimise string arguments.
+    // If the peeked-at compile time argument is something other than
+    // a string, assume it won't be a string later either.
+    let comptime_arg = jit_peek_at_stack(jit, ctx, 0);
+    if ! unsafe { RB_TYPE_P(comptime_arg, RUBY_T_STRING) } {
         return false;
     }
 
     // Generate a side exit
     let side_exit = get_side_exit(jit, ocb, ctx);
 
     // Guard that the argument is of class String at runtime.
-    let arg_opnd = ctx.stack_opnd(0);
-    mov(cb, REG0, arg_opnd);
-    jit_guard_known_klass(
-        jit,
-        ctx,
-        cb,
-        ocb,
-        unsafe { rb_cString },
-        StackOpnd(0),
-        comptime_arg,
-        SEND_MAX_DEPTH,
-        side_exit,
-    );
+    let comptime_arg_type = ctx.get_opnd_type(StackOpnd(0));
+    if comptime_arg_type != Type::CString && comptime_arg_type != Type::TString {
+        let insn_opnd = StackOpnd(0);
+        if !comptime_arg_type.is_heap() {
+            add_comment(cb, "guard not immediate");
+            test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK as i64));
+            jnz_ptr(cb, side_exit);
+            cmp(cb, REG0, imm_opnd(Qnil.into()));
+            jbe_ptr(cb, side_exit);
+
+            ctx.upgrade_opnd_type(insn_opnd, Type::UnknownHeap);
+        }
+        let arg_opnd = ctx.stack_opnd(0);
+        mov(cb, REG0, arg_opnd);
+        guard_object_is_string(cb, REG0, REG1, side_exit);
+        // We know this has type T_STRING, but not necessarily that it's a ::String
+        ctx.upgrade_opnd_type(insn_opnd, Type::TString);
+    }
```

Comment:
This seems a bit weird because presumably, even if the comptime argument is a string, you might still need to have guards/checks in place, because that doesn't guarantee all future arguments will be strings?

Maybe you could construct a test where you first pass something to a method with a shovel inside that is a string, and then you pass an integer to the same method to make sure things don't explode.

Reply:
That's actually the segfault I'm seeing - Alan suggested that might be a problem, and it is. Working on it.

Reply:
Though presumably the second segfault is different.

Reply:
You might already be on this but, I would suggest creating some adversarial tests to try and reproduce the segfault(s). Worth testing concat more extensively since we're trying to specialize on it.

Reply:
I have the beginning of one, yeah. Right now it just checks plus and shovel, but would be easy to expand. For starters we have a *lot* of code in test_yjit.rb that would benefit from the same checking.

Basically: for a lot of different input items (e.g. integer, various empty/small strings, float number, etc) pass them as the first arg to an operation to "bake" it in. Then try *every* item on that list as the value on following calls. Make sure the the results don't segfault, and that they have the same result as without YJIT.

A lot of things get exceptions, of course. I just return those and inspect them. For basically all operations, an exception will be raised rather than returned.

This *could* be a unit test, but if we add many operations it'll get slow enough that we don't want it in GitHub Actions. So I haven't structured it that way.

But it's not far along, and our ARM backend is *really* not ready for it, so right now I just have it as a test for some of my stuff.

Reply:
We're not testing the arm backend on `test_yjit.rb` yet.

Don't worry about adding too many tests. The bootstraptest runner can run tests in parallel now, and our tests run relatively fast. As long as you don't have long loops or really massive amounts of code in there, it will be fine.

Reply:
Is `comptime_arg_type` a typo here? `ctx.get_opnd_type` should be returning the type we _know_ not just what we saw at compile time.

Reply:
Aaah, if this comes from `ctx.get_opnd_type` then `comptime_arg_type` is indeed a poorly chosen name.